### PR TITLE
DEV-2053_account_verification

### PR DESCRIPTION
**Pull request by njharman**
#### What's this PR do?
Add verification endpoint for users to validate email. and API endpoints to trigger resending verification email.

  - AC1.c.i 200 with JSON {"detail": "Success"}
  - AC1.c.ii 404 with JSON {"detail": "Account already verified"}
  - added new AC1.c.iii 404 with JSON {"detail": "Account inactive"}

AC2.c SPA URL as undefined and implemented AC2.b as redirect to SPA instead of custom html template. These SPA routes need to be implemented in DEV-2027

    - "/verified/"           good to go!
    - "/verified/failed/" for reason we are not revealing (security)
    - "/verified/expired/"   took too long to follow verification link
    - "/verified/inactive/"  user.is_active == False
    - "/verified/unknown/"   user matching given email not found (maybe it got changed?)

Note: It should not be possible to malactor to generate verification links with arbitrary email addresses to fish our database.

Verification links expire after 24 hours (configurable in settings, can be disabled). 

Also,
  - Cut commented import that black/isort had different opinions on.
  - Fix error in README.md and duplication in settings/local.example.py.
  - Move create_test_user to factory.

#### Why are we doing this? How does it help us?
Unblock [DEV-2027]

#### How should this be manually tested? Please include detailed step-by-step instructions.
I believe the testing would be done in DEV-2027 (front end component)

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?
No

#### Have automated unit tests been added? If not, why?
Yes

#### How should this change be communicated to end users?
N/A

#### Are there any smells or added technical debt to note?
No, well maybe the SPA urlstructure detailed above not being "how SPA does things". And exposing inactive/unknown in url ?

#### Has this been documented? If so, where?
Maybe wait until front end done, but explaining email address must be verified for new accounts, will get email with link, the verification link expires (currently after 24 hours (subject to change))

If email is updated, new verification email will have to be made. 

#### What are the relevant tickets? Add a link to any relevant ones.
[DEV-2053]

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

Changed envvar/setting UID_SALT renamed to ENCRYPTION_SALT. 

New setting ACCOUNT_VERIFICATION_LINK_EXPIRY if non-zero will expire
account verification links after that many hours. If "0" or None links never expire. Default to 24 hours.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:
Yes [DEV-2027]

[DEV-2027]: https://news-revenue-hub.atlassian.net/browse/DEV-2027?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/tests/test_views.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ class CreateStripePaymentErrorConditionsTest(StripePaymentViewTestAbstract):
     """Branch coverage for various errors."""
 
     @override_settings(CELERY_TASK_ALWAYS_EAGER=True)
-    def atest_intervals(self):
+    def atest_intervals(self):  # TODO: njharman forgot to reenable/fix this test
```
Comment (x110dc) [inline apps/contributions/tests/test_views.py:114]: What did njharman do? 😉

File: revengine/settings/base.py [hunk 2]
```diff
@@ -450,9 +450,12 @@
 # across origins. Once this API supports public access, this needs to be loosened.
 AUTH_COOKIE_SAMESITE = "Strict"  # or 'Lax' or None
 
-# this is used as a salt for the UID hash.
+# Used as a salt for the UID hash and account verification links.
 UID_SALT = os.getenv("UID_SALT", "")
```
Comment (x110dc) [inline revengine/settings/base.py:454]: If it's going to be used for multiple purposes I think I'd prefer that it be renamed to something more generic.

**Review by x110dc — CHANGES_REQUESTED** [review]
Didn't review business logic, I'll leave that to @benjaminEwhite. Just a couple of comments/requests.

File: apps/users/tests/test_views.py [hunk 3]
```diff
@@ -441,3 +557,42 @@ def test_delete_not_implemented(self):
...
+    def test_request_account_verification_requires_auth(self):
+        # Requires Authenticated User.
+        response = self.client.get(reverse("user-request-account-verification"))
+        self.assertEqual(response.status_code, 401)
+
+    @patch("apps.users.views.send_templated_email")
+    def test_send_verification_email(self, sent):
+        user = create_test_user(is_active=True, email_verified=False)
+        UserViewset.send_verification_email(user)
+        assert sent.delay.called or sent.called
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/users/tests/test_views.py:591]: Would be better to not stub out send email functionality, and instead to inspect the sent email and make assertions about it.  Can see an example of that here: https://github.com/newsrevenuehub/rev-engine/pull/616/files#diff-fced53f883ab4297c7fd07e9170237ecf4a1c5e4378c0d91000253e5cf520b6fR115

As is, we don't have any tests that prove that the email that gets sent to users behaves as expected.

[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/users/tests/test_views.py:591]: Also, this test should be happening via the API layer, not calling the send_verification_email method on the viewset directly, IMO.

**Review by benjaminEwhite — CHANGES_REQUESTED** [review]
Left a bunch of comments.

I ran locally, changed `send_templated_email.delay` to `send_templated_email` in the `.send_verification_email` method so I could see the email that gets sent in log.  Looks like the verification URL is not getting converted in the email:

Screen Shot 2022-08-29 at 1 30 33 PM

File: revengine/settings/base.py (hunk 2 above)
Comment (njharman) [reply revengine/settings/base.py:454]: changed / documented to ENCRYPTION_SALT

File: apps/contributions/tests/test_views.py (hunk 1 above)
Comment (njharman) [reply apps/contributions/tests/test_views.py:114]: created DEV-2295 to resolve

File: apps/api/authentication.py [hunk 4]
```diff
@@ -41,6 +41,7 @@ def validate_token(self, token):
         if isinstance(user, Contributor) and not self.token_is_contrib_long_token(validated_token):
             return None, None
 
+        # Validate users that don't have is_active, or, that do and it's set True.
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/api/authentication.py:44]: I added comment after "simplyfying" logic line and breaking things. But, I'll accept that is failure of reader and not lack of clarity in code.  Cutting comment

File: apps/contributions/tests/test_views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (njharman) [reply apps/contributions/tests/test_views.py:114]: I'm not. Already added ticket.

File: apps/users/tests/test_views.py [hunk 5]
```diff
@@ -25,14 +28,127 @@
...
+from apps.users.tests.factories import create_test_user
+from apps.users.views import AccountVerification, UserViewset
 
 
 user_model = get_user_model()
 fake = Faker()
 
 
+class TestAccountVerificationEndpoint(TestCase):
+    @override_settings(ACCOUNT_VERIFICATION_LINK_EXPIRY=None)
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:40]: test should not assume/be dependent on state of world.  This setting change should not be able to cause test to start failing (or change which branches it covers)

Also I did (or will if I forgot) add test with / without expiry. As they'll exercise different code paths.

File: apps/users/tests/test_views.py [hunk 6]
```diff
@@ -25,14 +28,127 @@
...
+    def test_happy_path(self):
+        user = create_test_user(is_active=True, email_verified=False)
+        email, token = AccountVerification().generate_token(user.email)
+        response = self.client.get(reverse("account_verification", kwargs={"email": email, "token": token}))
+        self.assertRedirects(response, reverse("spa_account_verification"))
+        assert get_user_model().objects.get(pk=user.id).email_verified
+
+    @override_settings(ACCOUNT_VERIFICATION_LINK_EXPIRY=None)
+    def test_failed(self):
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:49]: I added more tests. But still won't like my terse names test_unknown_user, test_inactive_user, test_failed_bad_token, test_expired_token   [the first word after test_ corresponds to url]

File: apps/users/tests/test_views.py [hunk 7]
```diff
@@ -25,14 +28,127 @@
...
+        email, token = AccountVerification().generate_token("bobjohnny@example.com")
+        response = self.client.get(reverse("account_verification", kwargs={"email": email, "token": token}))
+        self.assertRedirects(response, reverse("spa_account_verification_fail", kwargs={"failure": "unknown"}))
+
+
+class TestAccountVerification:
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:55]: I did go back and add view tests for all but expiration, which is pain (all those tests vary only by one word). Originally the view tests were just to get coverage % over two branches. But there is little value in better testing there.

Disagree on where logic should be tested though.

I'm pretty "pure" on unit tests should test the unit and not be integration tests. [attenuated by pragmatism and lack of integration testing framework better than pytest]  I put all the logic in a separate class and not the view, to make it simple and direct to test. And to separate logic layer from view layer. Ideally views should be devoid of "logic" and only concern with display. rendering, encoding, protocol stuff like status code.

You don't test that django.orm.select works cause you expect that to be tested in django.orm. Likewise you don't test the AccountVerification works (when you use it) cause you expect that ito be tested in AccountVerification.

in my opinion anyways ;)

btw in other projects we've used pytest to explicitly write integration tests. they were kept separate and run separately (to keep unittests fast).

File: apps/users/tests/test_views.py [hunk 8]
```diff
@@ -25,14 +28,127 @@
...
+    )
+    def test_encoding_decoding(self, plaintext):
+        t = AccountVerification()
+        encoded = t.encode(plaintext)
+        decoded = t.decode(encoded)
+        assert str(plaintext) != encoded
+        assert str(plaintext) == decoded
+        assert re.match(
+            r"^[=a-zA-Z0-9._~-]*$", encoded
+        )  # RFC-3986 URL Safe, plus "=", equal, cause base64 says URL doesn't include query portion.
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:149]: The regex? Really? It's a single regex element. The group "[]" of those characters, zero or more times "*", bounded by start "^" and end "$" of string. I don't know how to write it more verbosely even with regex whitespace (which I loath) it'd be "~ [=a-zA-Z0-9._~-]* $" or with spaces between each character class in group.

The comment pointing to RFC explains intention better than any comment restating the regex in english would. 
I'll update to "Ensure string contains only RFC-3986 URL Safe characters plus equalsign, "=". Because base64 lib uses equalsigns in violation of RFC."

File: apps/users/tests/test_views.py [hunk 9]
```diff
@@ -441,3 +557,42 @@ def test_delete_not_implemented(self):
...
+        response = self.client.get(reverse("user-request-account-verification"))
+        assert status.HTTP_200_OK == response.status_code
+        assert {"detail": "Account already verified"} == response.json()
+        # Inactive
+        user.email_verified = False
+        user.is_active = False
+        user.save()
+        response = self.client.get(reverse("user-request-account-verification"))
+        assert status.HTTP_200_OK == response.status_code
+        assert {"detail": "Account inactive"} == response.json()
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:580]: No laziness under Ben's watch!  :)

Splitting up like I should have initially.

File: apps/users/tests/test_views.py [hunk 10]
```diff
@@ -441,3 +557,42 @@ def test_delete_not_implemented(self):
...
+
+    def test_request_account_verification_requires_auth(self):
+        # Requires Authenticated User.
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:583]: agree

File: apps/users/tests/test_views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:591]: > Why the or? 
I'm not testing how (sync or async) it's called. Don't want to test to fail if that changes.

Updated test to check email contents.

File: apps/users/views.py [hunk 11]
```diff
@@ -43,6 +50,79 @@
...
+def account_verification(request, email, token):
+    """Endpoint for verification link in "verify your email address" emails."""
+    checker = AccountVerification()
+    if user := checker.validate(email, token):
+        user.email_verified = True
+        user.save()
+        return HttpResponseRedirect(reverse("spa_account_verification"))
+    else:
+        # Having fail in url is non-optimal
+        return HttpResponseRedirect(reverse("spa_account_verification_fail", kwargs={"failure": checker.fail_reason}))
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:63]: per AC2 This is non-API endpoint. It's the link users click in email.  The endpoints that SPA uses to trigger sending email are in UserViewset.request_account_verification

File: apps/users/urls.py [hunk 12]
```diff
@@ -13,6 +13,7 @@
 ]
 
 orgadmin_user_management_urls = [
+    path("verify-account/<email>/<token>/", views.account_verification, name="account_verification"),
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/urls.py:16]: per AC2 This is non-API endpoint. It's the link users click in email. The endpoints that SPA uses to trigger sending email are in UserViewset.request_account_verification

File: apps/users/views.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:63]: My understanding DRF is API only. No?

File: apps/users/views.py [hunk 13]
```diff
@@ -43,6 +50,79 @@
...
+
+
+class AccountVerification(signing.TimestampSigner):
+    def __init__(self):
+        self.fail_reason = "failed"  # Single word [inactive, expired, unknown], other failures are empty string.
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:68]: Rewrote and explained.

File: apps/users/views.py [hunk 14]
```diff
@@ -43,6 +50,79 @@
...
+        self.max_age = settings.ACCOUNT_VERIFICATION_LINK_EXPIRY
+        super().__init__(salt=settings.UID_SALT)
+
+    def _hash(self, plaintext):
+        return get_sha256_hash(plaintext)
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:73]: It's used in 3 places, one far from others. Don't want someone to change one place and not others. [Sort expected get_sha..() to not work cause the truncation it does and expected to have to change it.]

File: apps/users/views.py [hunk 15]
```diff
@@ -43,6 +50,79 @@
...
+
+    def generate_token(self, email):
+        encoded_email = self.encode(email)
+        if self.max_age is None:
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:77]: AC didn't mention having expiration. So don't want to force that implementation (and I don't want to deal with discussing it with stakeholders).  But I agree there is no use case for not expiring. Why I made settings default to have one.

I agree with test to require settings be set (in general) [I desperately want to add test that fails if ENCRYPTION_SALT is empty string]

I did add expiry and no expiry test paths (to endpoint tests).

File: apps/users/views.py [hunk 16]
```diff
@@ -43,6 +50,79 @@
...
+        email = self.decode(encoded_email)
+        token = self.decode(encoded_token)
+        if not (email and token):
+            logger.info("Account Verification: Malformed or missing email/token for email:%s", email)
+            return False
+        if self.max_age:
+            try:
+                token = self.unsign(token, self.max_age)
+            except signing.SignatureExpired:
+                logger.warning("Account Verification: URL Expired for email:%s", email)
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:94]: yes, updated to include space all over.

File: apps/users/views.py [hunk 17]
```diff
@@ -43,6 +50,79 @@
...
+                self.fail_reason = "expired"
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:95]: This was poor decision on my part (having default be "failed") I've changed to have default be empty string and explicitly  set "failed" in the other previously dependent on default branches.

"failed" cause we don't want to leak info.  I would want to use "failed" for "unknown" but as stated in commit/PR comments I don't believe malactor should be able to generate a arbitrary links to probe for emails.  "inactive" proly should be failed too. But none of this was explored/defined in AC. "expired" seems like needful information to convey to good but slow user.

File: apps/users/views.py [hunk 18]
```diff
@@ -43,6 +50,79 @@
...
+                return False
+            except signing.BadSignature:
+                logger.info("Account Verification: Bad Signature for email:%s", email)
+                return False
+        if token != self._hash(email):
+            logger.info("Account Verification: Invalid token for email:%s", email)
+            return False
+        for user in get_user_model().objects.filter(email=email):
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:103]: Yes. I copied this from somewhere, and looking at it now, it's too goofy. changing.

File: apps/users/views.py [hunk 19]
```diff
@@ -89,19 +169,35 @@ class UserViewset(
...
+            "request_account_verification": [
+                IsAuthenticated,
+            ],
+        }.get(self.action, [])
         return [permission() for permission in permission_classes]
 
+    @staticmethod
+    def send_verification_email(user):
+        """Send email to user asking them to click verify their email address link."""
+        if not user.email:
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:189]: It can be empty string. If method passed User that hasn't been saved to db or validated it can be anything.

Can catch and log (these, admittedly, edge cases) or let django send_mail to raise or fail silently (no idea what it does)?

eh?

File: apps/users/views.py [hunk 20]
```diff
@@ -89,19 +169,35 @@ class UserViewset(
...
+            logger.warning("Account Verification: No email for user:%s", user.id)
+            return
+        encoded_email, token = AccountVerification().generate_token(user.email)
+        send_templated_email.delay(
+            user.email,
+            EMAIL_VERIFICATION_EMAIL_SUBJECT,
+            "nrh-org-account-creation-verification-email.txt",
+            "nrh-org-account-creation-verification-email.html",
+            {"verification_url": reverse("account_verification", kwargs={"email": encoded_email, "token": token})},
```
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/views.py:198]: done

File: revengine/settings/base.py [hunk 21]
```diff
@@ -450,9 +450,12 @@
...
 
+# Expire account verification URLs after X seconds.
+ACCOUNT_VERIFICATION_LINK_EXPIRY = 60 * 60 * 4  # Four hours
```
[the remark this answers is not in the document]
Comment (njharman) [reply revengine/settings/base.py:457]: I did consider! But 1) I don't think it will change often (or per environment, or need to be secret (the two reasons I'd put something in envvar)) 2) wasn't even an AC so reluctant to add more code/documentation/work for it.

i did change it to be in hours and not seconds to be more consistent with something I think I saw you working on.  
Maybe this? DJANGO_REST_MULTITOKENAUTH_RESET_TOKEN_EXPIRY_TIME = 24

I'm changing four hours to 24 to match that

File: revengine/urls.py [hunk 22]
```diff
@@ -69,8 +69,10 @@
     urlpatterns.append(path("dummy-500-error", dummy_view_for_raising_500, name="dummy-500"))
 
 urlpatterns += [
-    path(r"cloudflare-500", cloudflare_500_view, name="cloudflare-500"),
+    path("cloudflare-500", cloudflare_500_view, name="cloudflare-500"),
     # React SPA:
-    path(r"", index, name="index"),  # for reverse()
+    path("", index, name="index"),  # for reverse()
+    path("verified", index, name="spa_account_verification"),  # for reverse()
+    path("verified/<failure>/", index, name="spa_account_verification_fail"),  # for reverse()
```
[the remark this answers is not in the document]
Comment (njharman) [reply revengine/urls.py:76]: Use reverse() anywhere. I was just copying existing comment. But it took me while to understand what it meant...

Will update to "Routing handled by SPA, urlpatten defined here only to allow use of reverse()"

File: apps/users/tests/test_views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (njharman) [reply apps/users/tests/test_views.py:149]: This one is more complex, no? Uses (?:) group thing and has two elements?
 
> re_path(r"^(?:.*)/?$", ...

**Review by benjaminEwhite — APPROVED** [review]
LGTM!

**Review by x110dc — APPROVED** [review]
